File: frontend/src/game/statEffects.ts
import type { CharacterStats } from './types';

const clamp = (v: number, min = 0, max = 100) => Math.max(min, Math.min(max, v));

export function applyEffects(stats: CharacterStats, effects: Record<string, number>) {
  for (const [key, value] of Object.entries(effects)) {
    switch (key.toLowerCase()) {
      case 'money':
        stats.money = clamp(stats.money + value, 0, 999999);
        break;
      case 'mood':
        stats.mood = clamp(stats.mood + value);
        break;
      case 'energy':
        stats.energy = clamp(stats.energy + value);
        break;
      case 'health':
        stats.health = clamp(stats.health + value);
        break;
      case 'reputation':
        stats.reputation = clamp(stats.reputation + value);
        break;
      case 'stress':
        stats.stress = clamp(stats.stress + value);
        break;
      case 'awareness':
        stats.awareness = clamp(stats.awareness + value);
        break;
      case 'sensitivity':
        stats.sensitivity = clamp(stats.sensitivity + value);
        break;
      case 'intuition':
        stats.intuition = clamp(stats.intuition + value);
        break;
      case 'underworldbond':
        stats.underworldBond = clamp(stats.underworldBond + value);
        break;
      case 'influenceresistance':
        stats.influenceResistance = clamp(stats.influenceResistance + value);
        break;
      case 'willpower':
        stats.willpower = clamp(stats.willpower + value);
        break;
    }
  }
}

export function clampStats(stats: CharacterStats) {
  stats.mood = clamp(stats.mood);
  stats.energy = clamp(stats.energy);
  stats.health = clamp(stats.health);
  stats.reputation = clamp(stats.reputation);
  stats.stress = clamp(stats.stress);
  stats.awareness = clamp(stats.awareness);
}

File: backend/src/PeopleOfDarkMind.Application/Services/StatEffects.cs
using System.Text.Json;
using PeopleOfDarkMind.Domain.Entities;

namespace PeopleOfDarkMind.Application.Services;

public static class StatEffects
{
    public static void Apply(CharacterStats stats, string effectsJson)
    {
        if (string.IsNullOrWhiteSpace(effectsJson) || effectsJson == "{}") return;

        try
        {
            var effects = JsonSerializer.Deserialize<Dictionary<string, int>>(effectsJson);
            if (effects is null) return;

            foreach (var (key, value) in effects)
            {
                switch (key.ToLowerInvariant())
                {
                    case "money": stats.Money = Clamp(stats.Money + value, 0, 999999); break;
                    case "mood": stats.Mood = Clamp(stats.Mood + value); break;
                    case "energy": stats.Energy = Clamp(stats.Energy + value); break;
                    case "health": stats.Health = Clamp(stats.Health + value); break;
                    case "reputation": stats.Reputation = Clamp(stats.Reputation + value); break;
                    case "stress": stats.Stress = Clamp(stats.Stress + value); break;
                    case "awareness": stats.Awareness = Clamp(stats.Awareness + value); break;
                    case "sensitivity": stats.Sensitivity = Clamp(stats.Sensitivity + value); break;
                    case "intuition": stats.Intuition = Clamp(stats.Intuition + value); break;
                    case "underworldbond": stats.UnderworldBond = Clamp(stats.UnderworldBond + value); break;
                    case "influenceresistance": stats.InfluenceResistance = Clamp(stats.InfluenceResistance + value); break;
                    case "willpower": stats.Willpower = Clamp(stats.Willpower + value); break;
                }
            }
        }
        catch
        {
            // ignore malformed json
        }
    }

    private static int Clamp(int value, int min = 0, int max = 100) =>
        Math.Max(min, Math.Min(max, value));
}
